@here/harp-mapview/lib/geometry/TileGeometryLoader.ts

```diff
@@ -225,32 +242,46 @@ export class TileGeometryLoader {
      * Dispose of any resources.
      */
     dispose(): void {
-        this.reset();
+        this.clear();
         this.m_state = TileGeometryLoaderState.Disposed;
+        this.m_rejectFinishedPromise?.();
     }
 
     /**
      * Reset the loader to its initial state and cancels any asynchronous work.
      */
     reset(): void {
+        this.clear();
+
+        if (
+            this.m_state === TileGeometryLoaderState.Finished ||
+            this.m_state === TileGeometryLoaderState.Canceled ||
+            this.m_state === TileGeometryLoaderState.Disposed
+        ) {
+            this.m_finishedPromise = new Promise((resolve, reject) => {
+                this.m_resolveFinishedPromise = resolve;
+                this.m_rejectFinishedPromise = reject;
+            });
+        }
+        this.m_state = TileGeometryLoaderState.Initialized;
+    }
+
+    private clear() {
         this.m_availableGeometryKinds?.clear();
         this.m_enabledKinds?.clear();
         this.m_disabledKinds?.clear();
-
         this.m_decodedTile = undefined;
-        this.m_state = TileGeometryLoaderState.Initialized;
     }
 
     private finish() {
-        this.m_tile.loadingFinished();
-        this.m_tile.removeDecodedTile();
-
         this.m_decodedTile = undefined;
         this.m_state = TileGeometryLoaderState.Finished;
+        this.m_resolveFinishedPromise?.();
     }
 
     private cancel() {
         this.m_state = TileGeometryLoaderState.Canceled;
+        this.m_rejectFinishedPromise?.();
```

Comment:
Also, can you create a test for the state canceled and check that this promise is called?

Reply:
This is already covered in two tests: 
"should not start geometry creation for invisible tile"
"should not start geometry loading for invisible tile"

Note that cancel is private. It's called indirectly whenever the update() method is called for an invisible tile.